### ISSUE TITLE
Bleadperl v5.21.0-274-ga7ab896 breaks SHURIKO/String-Simrank-0.079.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13912]
Migrated from [rt.perl.org#122077](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122077) (status was 'rejected')

Searchable as RT122077$

**Comment by p5pRT** [comment]
### From @andk

git bisect


---
commit a7ab896004fe7cc32eeddadf760d0829e9fed13d
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Thu Jun 5 15&#8203;:03&#8203;:32 2014 \+0100

&nbsp;    when unCOWing a string\, set SvCUR to 0

diagnostics


---
http&#8203;://www\.cpantesters\.org/cpan/report/ccebd4e2\-ed7b\-11e3\-99f8\-9a0573305bfd

perl \-V


---
Summary of my perl5 \(revision 5 version 21 subversion 1\) configuration&#8203;:
&nbsp;  Commit id&#8203;: b823713ce3ee6bfd4d009e6307703c7d2e63b7c8
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.13\-1\-amd64\, archname=x86\_64\-linux\-ld
&nbsp;    uname='linux k83 3\.13\-1\-amd64 \#1 smp debian 3\.13\.7\-1 \(2014\-03\-25\) x86\_64 gnulinux '
&nbsp;    config\_args='\-Dprefix=/home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-275\-gb823713/127e \-Dmyhostname=k83 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Uuseithreads \-Duselongdouble \-DDEBUGGING=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=define
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='long double'\, nvsize=16\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=16\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/4\.8/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.18\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.18'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_NEW\_COPY\_ON\_WRITE PERL\_PRESERVE\_IVUV
&nbsp;                        PERL\_USE\_DEVEL USE\_64\_BIT\_ALL USE\_64\_BIT\_INT
&nbsp;                        USE\_LARGE\_FILES USE\_LOCALE USE\_LOCALE\_COLLATE
&nbsp;                        USE\_LOCALE\_CTYPE USE\_LOCALE\_NUMERIC USE\_LONG\_DOUBLE
&nbsp;                        USE\_PERLIO USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at Jun  5 2014 18&#8203;:21&#8203;:29
&nbsp;  %ENV&#8203;:
&nbsp;    PERL5LIB=""
&nbsp;    PERL5OPT=""
&nbsp;    PERL5\_CPANPLUS\_IS\_RUNNING="9004"
&nbsp;    PERL5\_CPAN\_IS\_RUNNING="9004"
&nbsp;    PERL\_MM\_USE\_DEFAULT="1"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-275\-gb823713/127e/lib/site\_perl/5\.21\.1/x86\_64\-linux\-ld
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-275\-gb823713/127e/lib/site\_perl/5\.21\.1
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-275\-gb823713/127e/lib/5\.21\.1/x86\_64\-linux\-ld
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.21\.0\-275\-gb823713/127e/lib/5\.21\.1
&nbsp;    \.
\-\- 
andreas

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Jun 10\, 2014 at 09&#8203;:28&#8203;:14PM \-0700\, Andreas J\. Koenig via RT wrote&#8203;:
> \# New Ticket Created by  \(Andreas J\. Koenig\) 
> \# Please include the string&#8203;:  \[perl \#122077\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122077 >
> 
> 
> git bisect
> \-\-\-\-\-\-\-\-\-\-
> commit a7ab896004fe7cc32eeddadf760d0829e9fed13d
> Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
> Date&#8203;:   Thu Jun 5 15&#8203;:03&#8203;:32 2014 \+0100
> 
>     when unCOWing a string\, set SvCUR to 0
> 
> diagnostics
> \-\-\-\-\-\-\-\-\-\-\-
> http&#8203;://www\.cpantesters\.org/cpan/report/ccebd4e2\-ed7b\-11e3\-99f8\-9a0573305bfd

It's a fault in the module \- its not safe under COW\.  The commit above
just made it so that the string in question was COW\, whereas previously it
had managed to dodge the bullet\.

Basically it was it was handling bit strings in XS and assuming it could
directly write to the PVX buffer\.

The following diff fixes it&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- ../../String-Simrank-0.079/lib/String/Simrank.pm	2010-07-27 20:03:42.000000000 +0100
+++ lib/String/Simrank.pm	2014-06-11 13:03:07.631713820 +0100
@@ -1182,7 +1182,6 @@
 
 /* Niels Larsen, May 2004. */
 
-static void* get_ptr( SV* obj ) { return SvPVX( SvRV( obj ) ); }
   /* A sub which takes as input the ScalarVector pointer.
      Returns a C string.
      The SvPVX function is part of the Perl internal API 
@@ -1190,6 +1189,12 @@
     SvRV dereferences a SV when its a perl reference
   */
 
+static void* get_ptr( SV* obj ) {
+    SV *str = SvRV(obj);
+    SvPV_force_nolen(str);
+    return SvPVX(str);
+}
+
 // define four macros for compiler to substitute where needed:
 #define DEF_SCOPTR( str )  int* scoptr = get_ptr( str )
     // receives a SV, returns scoptr: a pointer to an integer


-- 
```
</details>

"Foul and greedy Dwarf \- you have eaten the last candle\."
&nbsp;    \-\- "Hordes of the Things"\, BBC Radio\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Jun 11 05&#8203;:07&#8203;:56 2014\, davem wrote&#8203;:
> On Tue\, Jun 10\, 2014 at 09&#8203;:28&#8203;:14PM \-0700\, Andreas J\. Koenig via RT
> wrote&#8203;:
> > \# New Ticket Created by  \(Andreas J\. Koenig\)
> > \# Please include the string&#8203;:  \[perl \#122077\]
> > \# in the subject line of all future correspondence about this issue\.
> > \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122077 >
> >
> >
> > git bisect
> > \-\-\-\-\-\-\-\-\-\-
> > commit a7ab896004fe7cc32eeddadf760d0829e9fed13d
> > Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
> > Date&#8203;:   Thu Jun 5 15&#8203;:03&#8203;:32 2014 \+0100
> >
> > when unCOWing a string\, set SvCUR to 0
> >
> > diagnostics
> > \-\-\-\-\-\-\-\-\-\-\-
> > http&#8203;://www\.cpantesters\.org/cpan/report/ccebd4e2\-ed7b\-11e3\-99f8\-
> > 9a0573305bfd
> 
> It's a fault in the module \- its not safe under COW\.  The commit above
> just made it so that the string in question was COW\, whereas
> previously it
> had managed to dodge the bullet\.
> 
> Basically it was it was handling bit strings in XS and assuming it
> could
> directly write to the PVX buffer\.
> 
> The following diff fixes it&#8203;:
> 
> \-\-\- \.\./\.\./String\-Simrank\-0\.079/lib/String/Simrank\.pm    2010\-07\-27
> 20&#8203;:03&#8203;:42\.000000000 \+0100
> \+\+\+ lib/String/Simrank\.pm       2014\-06\-11 13&#8203;:03&#8203;:07\.631713820 \+0100
> @&#8203;@&#8203; \-1182\,7 \+1182\,6 @&#8203;@&#8203;
> 
> /\* Niels Larsen\, May 2004\. \*/
> 
> \-static void\* get\_ptr\( SV\* obj \) \{ return SvPVX\( SvRV\( obj \) \); \}
>    /\* A sub which takes as input the ScalarVector pointer\.
>       Returns a C string\.
>        The SvPVX function is part of the Perl internal API
> @&#8203;@&#8203; \-1190\,6 \+1189\,12 @&#8203;@&#8203;
>      SvRV dereferences a SV when its a perl reference
>    \*/
> 
> \+static void\* get\_ptr\( SV\* obj \) \{
> \+    SV \*str = SvRV\(obj\);
> \+    SvPV\_force\_nolen\(str\);
> \+    return SvPVX\(str\);
> \+\}
> \+
>  // define four macros for compiler to substitute where needed&#8203;:
>  \#define DEF\_SCOPTR\( str \)  int\* scoptr = get\_ptr\( str \)
>      // receives a SV\, returns scoptr&#8203;: a pointer to an integer

String\-Simrank is not a core distribution\.  Since Dave has kindly submitted his patch to String\-Simrank's bug queue \(https://rt.cpan.org/Public/Bug/Display.html?id=96377), we don't need to keep this RT ticket open\.  I will close the ticket and add a link\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @ikegami

Am I missing something\, or can

static void\* get\_ptr\( SV\* obj \) \{ SV \*str = SvRV\(obj\);
SvPV\_force\_nolen\(str\); return SvPVX\(str\); \}

be shortened to

static void\* get\_ptr\( SV\* obj \) \{ SV \*str = SvRV\(obj\); return
SvPV\_force\_nolen\(str\); \}



On Wed\, Jun 11\, 2014 at 8&#8203;:07 AM\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:

> On Tue\, Jun 10\, 2014 at 09&#8203;:28&#8203;:14PM \-0700\, Andreas J\. Koenig via RT wrote&#8203;:
> > \# New Ticket Created by  \(Andreas J\. Koenig\)
> > \# Please include the string&#8203;:  \[perl \#122077\]
> > \# in the subject line of all future correspondence about this issue\.
> > \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122077 >
> >
> >
> > git bisect
> > \-\-\-\-\-\-\-\-\-\-
> > commit a7ab896004fe7cc32eeddadf760d0829e9fed13d
> > Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
> > Date&#8203;:   Thu Jun 5 15&#8203;:03&#8203;:32 2014 \+0100
> >
> >     when unCOWing a string\, set SvCUR to 0
> >
> > diagnostics
> > \-\-\-\-\-\-\-\-\-\-\-
> >
> http&#8203;://www\.cpantesters\.org/cpan/report/ccebd4e2\-ed7b\-11e3\-99f8\-9a0573305bfd
>
> It's a fault in the module \- its not safe under COW\.  The commit above
> just made it so that the string in question was COW\, whereas previously it
> had managed to dodge the bullet\.
>
> Basically it was it was handling bit strings in XS and assuming it could
> directly write to the PVX buffer\.
>
> The following diff fixes it&#8203;:
>
> \-\-\- \.\./\.\./String\-Simrank\-0\.079/lib/String/Simrank\.pm    2010\-07\-27
> 20&#8203;:03&#8203;:42\.000000000 \+0100
> \+\+\+ lib/String/Simrank\.pm       2014\-06\-11 13&#8203;:03&#8203;:07\.631713820 \+0100
> @&#8203;@&#8203; \-1182\,7 \+1182\,6 @&#8203;@&#8203;
>
>  /\* Niels Larsen\, May 2004\. \*/
>
> \-static void\* get\_ptr\( SV\* obj \) \{ return SvPVX\( SvRV\( obj \) \); \}
>    /\* A sub which takes as input the ScalarVector pointer\.
>       Returns a C string\.
>       The SvPVX function is part of the Perl internal API
> @&#8203;@&#8203; \-1190\,6 \+1189\,12 @&#8203;@&#8203;
>      SvRV dereferences a SV when its a perl reference
>    \*/
>
> \+static void\* get\_ptr\( SV\* obj \) \{
> \+    SV \*str = SvRV\(obj\);
> \+    SvPV\_force\_nolen\(str\);
> \+    return SvPVX\(str\);
> \+\}
> \+
>  // define four macros for compiler to substitute where needed&#8203;:
>  \#define DEF\_SCOPTR\( str \)  int\* scoptr = get\_ptr\( str \)
>      // receives a SV\, returns scoptr&#8203;: a pointer to an integer
>
>
> \-\-
> "Foul and greedy Dwarf \- you have eaten the last candle\."
>     \-\- "Hordes of the Things"\, BBC Radio\.
>
